### ISSUE TITLE
Omit type annotations for local variables make code harder to understand.

### ISSUE THREAD

**Issue by bsutton** [dart-lang/sdk#39899]
The latest version of the 'default' google analyser settings emit the warning:

Omit type annotations for local variables (omit_local_variables_types).

In many cases this improves the readability of code by removing redundant information.

e.g.
```
var  i = 1;
```
The removal of the type (int) here doesn't affect the codes readability as the type of `i` is obvious.

There are however many cases where removing the type actually makes the code harder to read and for a human to do static analysis.

e.g.
```
    var code = _extractCode(monetaryAmount, codeLength);
```
There is no information on this line that tells a reader what type 'code' is.
The lack of type information forces the reader to stop and check what type _extractCode returns.

If you are analysing a class of some complexity, having to stop and check the type of numerous variables significantly slows down the process.

In a world of auto completion the reduction of a type to 'var' provides little reduction in typing.

The only other argument for reducing a type to 'var' is visual clutter.
Reduction in visual clutter should focus on removing redundant information, in the examples given the information being removed is not redundant.

I would suggest that the Omit type annotations warning should only be emitted when it is obvious what the type is.

**Comment by bwilkerson** [comment]
@davidmorgan

**Comment by eernstg** [comment]
Agreed, even on local variables a type annotation can amount to useful documentation, especially if it is initialized by a non-trivial expression involving entities declared in some other library.

For instance `var x = myComplex.expression(with, lots)..of.moving = part + s` may give `x` a type that is determined by looking up the return type of `expression` (which could be a generic method), filling in the results of type inference involving the static type of `with` and `lots`, and so on. The resulting type may not be immediately obvious, and the code gets _less maintainable_ if the type of `x` is subject to surprising and/or subtle changes because some declarations are modified in a library that some other organization maintains.

We might be able to use a syntactic criterion for determining which expressions have an "obvious" type, and only let the lint flag those local variables that are initialized by such expressions.

**Comment by bsutton** [comment]
When var was first introduced to Java I thought it was a fantastic idea. In practice the results were not very useful and just make code harder to read.

I really think var should be used sparingly.
It should only be used when the right hand side of an expression is one of:
A constructor
An intrinsic type

There is also an argument that changing int to var is pointless.

**Comment by davidmorgan** [comment]
Thanks everyone!

I personally handled the cleanup of google-internal code, so I have seen the effect of applying this lint to a very large body of code. It was overwhelmingly positive.

Although it's possible to come up with toy examples where type annotations add information this does not seem to happen in real code.

Notice that the same lints release requires return type annotations always, so types are fixed on API boundaries. Small methods were already a good way to improve readability; now they are also the main way to fix types if you feel that is necessary.

Nothing is set in stone and we will continue to tweak the recommended lints--including the implementations of the lints--if we see ways to improve them. Feedback is important in that process, so, thanks :)

**Comment by bsutton** [comment]
David, thanks for the response.

My experience is that it can make the code easier to read which I guess us
what you are referring to when you say that it was overwhelmingly positive.

My point is that whilst it may make it easier to causality read, it
definitely makes the code harder to analyze.


And if there is a trade off to be had then I would strongly vote for the
analysis to be easier.

I get by fine causality reading code with types.

**Comment by davidmorgan** [comment]
Thanks Brett. Yes, I meant it's easier to read without type annotations.

The most important use case we consider for interaction with code is when you are using the Dart analyzer. So, for example, browsing code in IntelliJ. Then you can easily discover types and navigate to them.

More important than the type, though, is the exact method you're calling. That gives you not just the type of the value, but the semantics of the value--what _this_ specific instance is supposed to do/be.

Or, with an example: why bother figuring out that you have a `List<String>`, when what you really want to know is part of the API of the method you're calling--that it's a list of names of special offers applicable to the logged in user.

Types are extremely powerful when it comes to catching errors and making sure code is doing something sensible/supported. But you mostly don't need to know about the details. The type system makes sure that the code fits together in a plausible way--but it can do that perfectly well without your help.

What you need to worry about as a developer is the _semantics_ of the information flowing about, which is a level the compiler can't help you with. The idea of this lint is that you should lean on the tools for the part that can be automated, leaving you to do what we still need a human for :)

**Comment by davidmorgan** [comment]
One more note, an appeal to the data :)

It looks like teams mostly either choose to never use type annotations or to always use them.

If it was the case that local type annotations sometimes add value, as argued, then there would be teams that sometimes use local type annotations--i.e., when they add value.

This does not seem to actually happen with any detectable frequency :)

**Comment by bsutton** [comment]
I agreed with the assumption that Devs should be using an ide and relying
on the analyser.

We happen to use visual code but that's not particularly important for this
conversation.

Whilst I agreed that you often need to dig into a method to understand why
a particular type is being returned, more often than not, the type
information is sufficient.

I also agree that we should let the tools do as much as possible.

So let me put this issue in context.

I'm an experinced dev with more the 20 years under my belt, working with a
team of similarly experienced devs. Most recently the team moved from Java
using eclipse and we have been eagerly awaiting a production upgrade of our
Java runtime so we can use var.

With an upgrad to dart 2.7 we went through and fixed all of the lint
warnings as we made the decision early on to follow the Google lint
defaults with pedantic turned on. We also don't allow code to be committed
with any lint warnings as they can obscure real problems.

Very shortly after committing the code complaints started flowing around
how it was harder to do deep analises which directly lead to this issue
being raised.

So my point is that issue isn't some esoteric whinge:), but rather a
problem encountered by experienced Devs hunting bugs in a large and active
code base.

It also isn't a result of a team that was some how anti using vars.

This change has made life harder when it really counts.

**Comment by bsutton** [comment]
I'm not certain how to interpret this data point and I'm somewhat concerned
that I will reject it out of hand simply because it didn't appear to
support my point of view.

With that said I should still raised some counter points.

I'm assuming this data is pre the new lint rule being applied otherwise
there would be no type annotations left in any of the code.

As such I wonder about the background of the teams and the local rules they
have applied.

There appears to be large chunks of Google libraries that are typeless (as
an aside these are the ones that are a pain to use). I'm guessing this is a
hang over from earlier versions of dart (was dart originally typeless?) Or
the Devs come from a JavaScript background.

So the question is, are the local team rules dictating this and that's why
you see no variance.

You would see no variance in our code base due to our team rules.

We know from experience that types are important.

Case in point.

We recently assisted in porting a JavaScript library to dart.
When we joined the team the port was mostly done but we were finding bugs
in the code base.

The code base had zero type annotations.
The code passed some thirty message types around.

When trying to analize the code we where getting no where. So we decided to
take the fairly invasive approach of adding types to core paths of the
library.

As you might expect the compiler spat out numerous errors that the original
JavaScript Devs weren't even aware of, but more importantly we were able to
reason about code in a way that simply wasn't possible pre adding types.

My point is, the further away the type information is the harder it is to
reason about the code base.

This particular lint is regressive because it's moving the type information
further from where the developer needs it.

**Comment by davidmorgan** [comment]
Thanks Brett--and in particular, thanks for giving more detail. It's always useful to hear about real life examples.

I'm a little puzzled about your example. Typically I expect all methods to have type annotations, and that this is easily sufficient to constrain the types and avoid the problems you describe.

In a case where methods don't all have type annotations, but some local variables do, I definitely agree that `omit_local_variable_types` is harmful. Adding type annotations to all methods signatures should be done before removing them from locals.

In a case where methods are overly long, and use explicitly typed local variables to break up the flow, I also think `omit_local_variable_types` is harmful. In this case, the methods should be split up before removing type annotations from locals.

If neither of these applies, then I don't really see where the problems you describe could be coming from?

Thanks again for the input :)
